### ISSUE TITLE
Provide support for running the Checker Framework

### ISSUE THREAD

**Issue by cushon** [bazelbuild/bazel#6356]
It would be nice if it was easy to enable the [Checker Framework](http://checkerframework.org) for `java_*` rules.

As an alternative to building the support directly in to the native rules, the integration could be implemented as an aspect that attaches to `Javac` and creates a parallel action that runs the Checker Framework's javac over the same sources.

One issue is that the Checker Framework doesn't yet support JDK 9 (see https://github.com/typetools/checker-framework/issues/1224), and Bazel is using JDK <strike>9</strike> <strike>10</strike> 11  as the default `--host_javabase`. We could work around that by running the Checker in the target configuration as a test, instead of in the host config as a validation action that breaks the build.

**Comment by the1derer** [comment]
@cushon @lberki  I will be working on [Nullness Checker Case Study on Bazel tool](https://summerofcode.withgoogle.com/projects/#5041382207520768) this Google Summer of Code where I would solve this issue 
My [proposal draft ](https://docs.google.com/document/d/1n7hrCNPD1qyqRyzScRaCn_JdVWwDQsMFpbKQPwstprc/edit?disco=AAAACrNJGz8) contains some initial work to implement this feature and enable projects to use Checker Framework while using bazel build tool.
Any additional information that I should know going forward is welcomed 
Thanks

**Comment by the1derer** [comment]
@cushon @laurentlb 
I am trying to enable the Checker Framework for `java_*` rules but has run into the following problems. It would be helpful if you could provide some insight into the problem.

**TL; DR: I have run into cyclic dependency while enabling Checker Framework. Just see https://github.com/the1derer/bazel-with-checkerframework/tree/Alternate-3 in "Alternate-3" branch Readme**

Since Checker-Framework only works with JDK8 and use a separate AnnotatedJDK8(and since Bazel itself uses embedded jdk11) I created the `java_toolchain` as shown below:-
Full Code is in https://github.com/the1derer/bazel-with-checkerframework/tree/Alternate-3 in "Alternate-3" branch.

I followed [How to build custom `java_toolchain`](https://groups.google.com/forum/#!topic/bazel-discuss/paNHmxWcY8Y) to configure custom javac and insert values into ijar, deploy, etc but I was not sure and didn't know what to put in `singlejar`

**BUILD file**
```starlark
java_library(
    name = "WithBug",
    srcs = ["src/java/com/example/WithBug.java"],
    plugins = ["checker-framework"],
    deps = [
        "@checker_qual//jar",
    ],
)

java_plugin(
    name = "checker-framework",
    processor_class = "org.checkerframework.checker.nullness.NullnessChecker",
    visibility = ["//visibility:public"],
    deps = [
        "@checker//jar",
    ],
)

java_toolchain(
    name = "annotated-toolchain",
    genclass = ["@bazel_tools//tools/jdk:genclass"],
    ijar = ["@bazel_tools//tools/jdk:ijar"],
    javabuilder = [
        "@bazel_tools//tools/jdk:javabuilder",
        "@bazel_tools//tools/jdk:vanillajavabuilder",
    ],
    bootclasspath = ["@annotatedjdk8//jar"], # Most probably this line is causing the cycle.
    javac = [":annotated-javac"],
    singlejar = ["cf/javac.jar"], # I am not sure what to put into this.
    source_version = "8",
    target_version = "8",
)

java_import(
    name = "annotated-jdk",
    jars = ["@annotatedjdk8"],
)

java_import(
    name = "annotated-javac",
    jars = ["cf/javac.jar"],
)
```
**WORKSPACE file**
```
maven_jar(
    name = "checker_qual",
    artifact = "org.checkerframework:checker-qual:2.8.1",
    sha1 = "eb2e8ab75598548cc8acf9a1ca227e480e01881e",
)

maven_jar(
    name = "checker",
    artifact = "org.checkerframework:checker:2.8.1",
    sha1 = "8b9453fbd9a1dff5aa1cad2c2ea4d0a8434b1db1",
)

maven_jar(
    name = "annotatedjdk8",
    artifact = "org.checkerframework:jdk8:2.8.1",
    sha1 = "99c6219b7688d0e01f3d6eab920392d4c1ce8d2d",
)

maven_jar(
    name = "javacutil",
    artifact = "org.checkerframework:javacutil:2.8.1",
    sha1 = "3b19bbffdefa1a59cfcc60b3291d0a2fe10ad4f8",
)

maven_jar(
    name = "dataflow",
    artifact = "org.checkerframework:dataflow:2.8.1",
    sha1 = "932060eb41297bc806ddf913530dbb51bb6929fb",
)
```

I ran the following commands separately:-
1. `bazel build --java_toolchain=:annotated-toolchain  :WithBug`
2. `bazel build --host_javabase=@local_jdk//:jdk  --java_toolchain=:annotated-toolchain :WithBug`  -----> JAVA_HOME is set at JDK8

was met with the same error everytime.
```
ERROR: /home/shubham/.cache/bazel/_bazel_shubham/b0f285083386ebbe43e61639b31a441d/external/bazel_tools/tools/jdk/BUILD:59:1: in java_toolchain_alias rule @bazel_tools//tools/jdk:current_java_
toolchain: cycle in dependency graph:
    //:WithBug
    @checker_qual//jar:jar
.-> @bazel_tools//tools/jdk:current_java_toolchain
|   @bazel_tools//tools/jdk:legacy_current_java_toolchain
|   //:annotated-toolchain
|   @annotatedjdk8//jar:jar
`-- @bazel_tools//tools/jdk:current_java_toolchain
This cycle occurred because of a configuration option
ERROR: Analysis of target '//:WithBug' failed; build aborted
INFO: Elapsed time: 3.850s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (15 packages loaded, 323 targets configured)
```

Can you help me in resolving this? Is this because bazel can't support JDK8 now?
I also tried replacing all maven-jars with local java_import but reached the same error.

My other Attempts/Alternative to integrate Checker-Framework with what errors I ran into is documented @ https://github.com/the1derer/bazel-with-checkerframework/blob/master/README.md

Thanks.

P.S. - My Ultimate Goal is to integrate Checker-Framework's Nullness Checker into Bazel's Source Code itself for my Google Summer of Code'19 Case Study.

**Comment by the1derer** [comment]
cc / @mernst @smillst

**Comment by cushon** [comment]
cc @iirina for the question about Bazel `java_toolchain` definitions

**Comment by regisd** [comment]
It seems some people have managed to run nullness checks using Uber's nullway.
https://stackoverflow.com/questions/48555320/run-checker-framework-with-bazel
